rectools/models/sasrec.py

```diff
@@ -0,0 +1,723 @@
+        self.device = torch.device(device)
+        self.n_threads = cpu_n_threads
+        self.model: TransformerBasedSessionEncoder
+        self._model = TransformerBasedSessionEncoder(
+            n_blocks=n_blocks,
+            factors=factors,
+            n_heads=n_heads,
+            session_maxlen=session_maxlen,
+            dropout_rate=dropout_rate,
+            use_pos_emb=use_pos_emb,
+            use_causal_attn=True,
+            transformer_blocks_type=transformer_blocks_type,
+            item_net_type=item_net_type,
+        )
+        self.trainer = Trainer(  # TODO: move to lightning trainer and add option to pass initialized trainer
+            lr=lr,
+            epochs=epochs,
+            device=self.device,
+            loss=loss,
+        )
+        self.data_preparator = SasRecDataPreparator(session_maxlen, batch_size)  # TODO: add data_preparator_type
+        self.u2i_dist = Distance.DOT
+        self.i2i_dist = Distance.COSINE
+
+    def _fit(
+        self,
+        dataset: Dataset,
+    ) -> None:
+        processed_dataset = self.data_preparator.process_dataset_train(dataset)
+        train_dataloader = self.data_preparator.get_dataloader_train(processed_dataset)
+
+        self.model = deepcopy(self._model)  # TODO: check that it works
+        self.model.costruct_item_net(processed_dataset)
+
+        self.trainer.fit(self.model, train_dataloader)
+        self.model = self.trainer.model
+
+    def _custom_transform_dataset_u2i(
+        self, dataset: Dataset, users: ExternalIds, on_unsupported_targets: ErrorBehaviour
+    ) -> Dataset:
+        filtered_dataset = self.data_preparator.process_dataset_recommend(dataset, users)
+        return filtered_dataset
+
+    def _custom_transform_dataset_i2i(
+        self, dataset: Dataset, target_items: ExternalIds, on_unsupported_targets: ErrorBehaviour
+    ) -> Dataset:
+        filtered_dataset = self.data_preparator.process_dataset_recommend_to_items(dataset)
+        return filtered_dataset
+
+    def _recommend_u2i(
+        self,
+        user_ids: InternalIdsArray,
+        dataset: Dataset,  # [n_rec_users x n_items + 1]
+        k: int,
+        filter_viewed: bool,
+        sorted_item_ids_to_recommend: tp.Optional[InternalIdsArray],  # model_internal
+    ) -> InternalRecoTriplet:
+
+        if sorted_item_ids_to_recommend is None:  # TODO: move to _get_sorted_item_ids_to_recommend
+            sorted_item_ids_to_recommend = self.data_preparator.get_known_items_sorted_internal_ids()  # model internal
+
+        self.model = self.model.eval()
+        self.model.to(self.device)
+
+        # Dataset has already been filtered and adapted to known item_id_map
+        recommend_dataloader = self.data_preparator.get_dataloader_recommend(dataset)
+
+        session_embs = []
+        item_embs = self.model.item_model.get_all_embeddings()  # [n_items + 1, factors]
+        with torch.no_grad():
+            for x_batch in tqdm.tqdm(recommend_dataloader):
```

Comment:
It's probably better to use `from tqdm.auto import tqdm` for pretty bars in jupyter. But it's not quite safe. Please add TODO here for the future

Reply:
fixed